### PR TITLE
MNTOR-983: Settings page frontend

### PR DESCRIPTION
**Pull request by flozia**
# References: 
Jira: [MNTOR-983](https://mozilla-hub.atlassian.net/browse/MNTOR-983)
Figma: [Monitor settings page](https://www.figma.com/file/gvyyqS9egsGjsH0LUDqT1m/Breach-Resolution-Flow-Design?node-id=214%3A20758&t=BXLEdrjwcip3gbdZ-0) (with changes to the [email list style](https://www.figma.com/file/gvyyqS9egsGjsH0LUDqT1m?node-id=214:20759#347392821))

# Description

⚠️ Depends on the backend changes introduced by PR #2773

This PR adds the frontend parts for the settings page.
# Screenshot

monitor-settings-page

# How to test

# Checklist (Definition of Done)
- [x] Localization strings (if needed) have been added.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [x] I've added or updated the relevant sections in readme and/or code comments
- [ ] I've added a unit test to test for potential regressions of this bug.
- [x] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [x] All acceptance criteria are met.
- [x] Jira ticket has been updated (if needed) to match changes made during the development process.
- [x] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: src/client/js/settings.js
Comment (flozia) [inline src/client/js/settings.js:82]: How should we best handle localized strings that are being set via JS — should we include `fluent`?

File: src/controllers/settings.js
Comment (flozia) [inline src/controllers/settings.js:7]: This data will be added by PR #2743 which introduces the backend changes for the settings page.

File: src/views/email-2022.js
Comment (flozia) [inline src/views/email-2022.js:1]: Email template taken from the previous implementation.

File: src/client/js/settings.js
Comment (rhelmer) [reply src/client/js/settings.js:82]: I think so, @toufali might have a better idea.

Comment (toufali) [reply src/client/js/settings.js:82]: The `getMessage()` function in utils/fluent.js accepts an optional argument that should help cases like this.  Example:

https://github.com/mozilla/blurts-server/blob/fe9e0c31aa18bb0968311abaa4791136fbaeb1eb/src/views/partials/breaches.js#L75

https://github.com/mozilla/blurts-server/blob/fe9e0c31aa18bb0968311abaa4791136fbaeb1eb/locales/en/breaches.ftl#L10-L16


Here's another (possibly less complex) example where the string is translated into a constant rather than template:
https://github.com/mozilla/blurts-server/blob/fe9e0c31aa18bb0968311abaa4791136fbaeb1eb/src/views/partials/breaches.js#L37-L42

https://github.com/mozilla/blurts-server/blob/fe9e0c31aa18bb0968311abaa4791136fbaeb1eb/locales/en/breaches.ftl#L31-L33

File: locales/en/settings.ftl
Comment (flodolo) [inline locales/en/settings.ftl:15]: This still needs a plural structure (like the one on line 22-27)

File: locales/en/settings.ftl
Comment (flodolo) [inline locales/en/settings.ftl:17]:
```suggestion
settings-resend-email-verification-link = Resend verification email
```

File: locales/en/settings.ftl
Comment (flodolo) [inline locales/en/settings.ftl:32]:
```suggestion
settings-deactivate-account-info = You can deactivate {-product-short-name} by deleting your {-brand-fx-account}.
```

File: locales/en/settings.ftl
Comment (flodolo) [inline locales/en/settings.ftl:33]:
```suggestion
settings-fxa-link-label = Go to {-brand-name} Settings
```

The naming of that term was pretty poor :-( (which brand?)

File: locales/en/settings.ftl
Comment (flodolo) [inline locales/en/settings.ftl:40]:
```suggestion
settings-send-email-verification-button = Send verification link
```

**Review by flodolo — CHANGES_REQUESTED** [review]
Thanks, the FTL file looks a lot better now, still a few things to fix.

File: src/client/js/settings.js
Comment (flozia) [reply src/client/js/settings.js:82]: @toufali Thanks for the examples. My question here specifically was more related to: If we set localized strings from `JS` we’d need to include `fluent.js` in the frontend resources or is there another way we could do this?

File: locales/en/settings.ftl
Comment (flozia) [reply locales/en/settings.ftl]: Just saw that we also have {[-brand-firefox](https://github.com/mozilla/blurts-server/blob/main/locales/en/app.ftl#L805)}. :)

File: src/client/js/settings.js
Comment (toufali) [reply src/client/js/settings.js:82]: Ah I see.  

I'm not familiar with using fluent.js on the client – wouldn't that require special endpoints or localization server, and might that introduce lag in showing a proper translation?  Would be interested to hear more of this approach.

The legacy site appears to creates the modal with translations on the server, and then pass additional translation strings as needed via XHR response after user interaction:

https://github.com/mozilla/blurts-server/blob/1d59e3383aed8d672dd2b130f1e30b940536f12f/public/js/resolve-breaches.js#L116-L157

Other options:
- use a data-attribute on the relevant tag (if it exists) on initial render
- include the translations in an SSR `<script>` tag (though I believe this pollutes global scope, so likely not a first choice without additional thought) :
```js
<script>
var l10n = {
         addEmailDialogContent: 'a translated string'
      }
</script>
```

- you could use an SSR `<template>` tag to store strings for use on client:
```html
<template>
  <p data-dialog-txt='translated string'></p>
  <!-- or -->
  <p data-dialog-txt>translated string</p>
</template>
```

- expanding on the previous, you could just use a template element for the entire structure you intend to render dynamically, not just translations...  This might play nicely with the lightbox/dialog web component forthcoming this sprint – i.e. the client web component is a shell/wrapper that loads the SSR template content 🤷

File: src/views/partials/settings.js
Comment (flozia) [reply src/views/partials/settings.js]: You are totally right — cleaned that up!

**Review by flodolo — APPROVED** [review]

**Comment by toufali** [comment]
I see a few references to https://github.com/mozilla/blurts-server/pull/2743 but that PR is closed?  Is there an updated PR, and should I review that one first?

**Comment by rhelmer** [comment]
> I see a few references to #2743 but that PR is closed? Is there an updated PR, and should I review that one first?

Sorry, I split this out to https://github.com/mozilla/blurts-server/pull/2773 (which is done except for the older tests that I am porting over and reviewing, we should land that and https://github.com/mozilla/blurts-server/pull/2772. I'll leave a comment in #2743 for posterity as well.

I'd suggest landing those first and then this current one (plus any small changes we need to properly pass the email verification template).

**Comment by toufali** [comment]
> I split this out to #2773 (which is done except for the older tests that I am porting over and reviewing), to support this PR we need to land that and #2772. I'll leave a comment in #2743 for posterity as well.
> 
> I'd suggest landing those first and then this current one (plus any small changes we need to properly pass the email verification template).

Thanks @rhelmer!  #2772 is approved and I'll keep my eye out for #2773 when it's ready for review.  After both are merged I can test this PR on top of them 👍

File: src/views/partials/settings.js
Comment (toufali) [inline src/views/partials/settings.js:132]: nit: since element IDs create global vars, should we avoid their use – at least on non-form elements?  If so, is this one avoidable?  The selector `.settings[data-csrf-token]` (or a similar variation) seems unique enough to me.

In the merged branch that I'm reviewing locally (prepared by @rhelmer) this ID is simply "settings", which is more of a concern as a global var.

File: src/views/partials/settings.js
Comment (toufali) [inline src/views/partials/settings.js:155]: See note about ID attribute global vars and if we should avoid for non-form elements.

File: src/views/partials/settings.js
Comment (toufali) [inline src/views/partials/settings.js:68]: Be careful here, I believe `checked` == `checked="true"` == `checked="false"`.  So, having `checked="false"` is actually specifying the checkbox _is_ checked.

You probably need to use a conditional here:
```suggestion
    ${isChecked ? 'checked' : ''}
```

https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes

File: src/client/css/partials/settings.css
Comment (toufali) [inline src/client/css/partials/settings.css:24]: Suggestion: make this page responsive across all viewports rather than just for mobile media-query.  The following style will wrap the settings page sooner, rather than waiting for the breakpoint:

```suggestion
.settings {
  display: flex;
  flex-wrap: wrap;
  gap: var(--padding-md) var(--padding-xl);
  padding: var(--padding-lg);
}

.settings-title,
.settings-section-title {
  color: var(--main-dark);
}

.settings-section-title {
  margin: var(--padding-sm) 0 var(--padding-md) 0;
}

.settings-content {
  color: var(--gray-40);
  font-size: 0.875rem;
  flex: 1 1 66%;
}
```

File: src/client/css/partials/settings.css
Comment (toufali) [inline src/client/css/partials/settings.css:158]: Per responsive comment above, this can be deleted if accepting suggestion:
```suggestion
```

**Review by toufali — CHANGES_REQUESTED** [review]
Good first pass!  I'm rushed for time in this review, but I wanted to post thoughts early since it's overdue:

I think a major blocker is that the Breach alert preferences radio buttons don't reflect the state I set them at – i.e. if I check the first radio labeled "Send breach alerts to the affected email address" and then reload, it still displays the second radio as selected ("Send all breach alerts to primary").  This could be just a result of `checkbox="false"` issue (see inline comment below) – I haven't gone deeper into the user db settings than this.

A minor blocker is responsiveness across viewports between mobile and desktop.  See screenshot and suggestion inline.

Screen Shot 2023-01-28 at 4 47 59 PM 

_Page at this viewport looks broken, and user is unable to scroll horizontally to see content._

File: locales/en/settings.ftl
Comment (flodolo) [inline locales/en/settings.ftl:1]: I haven't been strict about FTL files in this repository, but since this is a new file, I think it's worth cleaning up the formatting.

This is how the file would be formatted in mozilla-central:
* License header
* Spaces around message/term/variable references
* Mandatory comments for variables

Could you update this file accordingly? I'm attaching the updated FTL (renamed to .txt to be able to attach it here).

[settings.txt](https://github.com/mozilla/blurts-server/files/10529118/settings.txt)

I'd  like to enforce this style going forward for new strings.

File: src/client/js/settings.js
Comment (toufali) [inline src/client/js/settings.js:131]: Minor note, non-blocking: instead of using a conditional to check each interactive element on the page, you could simply verify that this is the settings partial.  Each "main" template is created with the partial name as a data attribute, so you could do:
```js
const settingsPartial = document.querySelector("[data-partial='settings']")
```
followed by a single conditional check similar to:
```js
if (settingsPartial) init()
```
Where `init()` is a function to set up listeners, run-once code, etc.

In addition to feeling confident the correct partial and elements are loaded, you have access to the `<main>` wrapper around your partial structure, and can query within it (`settingsPartial.getElementsByClassName('js-settings-resend-email'`) instead of the entire document.  This might allow less verbose names?

If interested, see https://github.com/mozilla/blurts-server/blob/main/src/client/js/partials/breaches.js for example.

File: locales/en/settings.ftl
Comment (flozia) [reply locales/en/settings.ftl:1]: Thank you @flodolo for putting together example FTL — updated and noted for any new additions.

File: src/client/css/variables.css
Comment (flozia) [reply src/client/css/variables.css:22]: Good callout. That probably sneaked it’s way in from the older styles. `Main Dark` [is being used in Figma](https://www.figma.com/file/gvyyqS9egsGjsH0LUDqT1m/Breach-Resolution-Flow-Design?node-id=272%3A17536&t=JKv9eigvfBIXIObU-4), but I’ll sync with UX and see if we can replace this.

File: src/client/js/settings.js
Comment (flozia) [reply src/client/js/settings.js:5]: Adding  `js` communicates clearly that this class is not used for styling, but as a target for scripting — separating those reduces the chances that the classname is used for other purposes than originally intended. Also I like the extra context that descriptive classnames give. If we have another convention or would like to introduce one I’m happy to adopt it.

File: src/views/partials/settings.js
Comment (flozia) [reply src/views/partials/settings.js:132]: I agree on trying to avoid using `id`s. For entire pages I think it might be helpful to have something more descriptive than `.settings[data-csrf-token]`. I’ll add an identifier to the class list.

File: src/client/css/partials/settings.css
Comment (flozia) [reply src/client/css/partials/settings.css]: Thanks for finding the layout issue and providing a great fix with this suggestion! :)

File: src/client/js/settings.js
Comment (flozia) [reply src/client/js/settings.js]: I like that suggestion — it also helps to grasp which dynamic elements are expected to be present easily.

**Comment by flozia** [comment]
> I think a major blocker is that the Breach alert preferences radio buttons don't reflect the state I set them at […]

Fixed by commit 8f98008 — I didn’t pass the users communication preferences to the partial at all.

**Review by rhelmer — APPROVED** [review]

File: src/client/js/settings.js
Comment (toufali) [reply src/client/js/settings.js]: Thanks for the context – I wasn't aware of that convention (which doesn't help either of our cases 😄 ).  

I've heard of using data-attributes for cases where an element needs to be selected but not styled, but I'm not sure if that's any better.  I usually resort to a specific-enough query selector on a stable parent and deal with any consequences later!

File: src/client/css/global.css
Comment (toufali) [inline src/client/css/global.css:183]: nit: I believe `ease` is the default value – we could likely delete this 🤷

File: src/client/css/partials/settings.css
Comment (toufali) [inline src/client/css/partials/settings.css:10]: I think this var doesn't exist any more?

**Review by toufali — APPROVED** [review]
This looks good!  Approving with a couple notes to follow up on:
- [I commented](https://github.com/mozilla/blurts-server/pull/2773#discussion_r1091442592) in @rhelmer 's PR that we need to update the images and associated paths for the verification email.  Not sure if we want to merge his PR and handle the update in yours since it's more on the "frontend" side, but I'll let y'all work that out.
- one last thought on all the DOM query + conditionals: I _think_ this settings page will make all those DOM queries for every page load, because all client scripts are bundled.  e.g. even though the Breaches page only queries the DOM once to ensure the right partial, it will also incur the cost of every DOM query and conditional made by Settings.  It might be a micro-optimization, but something to keep in mind – maybe we create a refactor ticket to optimize when we've got more time?

**Comment by flozia** [comment]
> This looks good! Approving with a couple notes to follow up on:

Thank you @toufali — I addressed your inline comments and regarding your other points:

- **Confirmation email images**
Just opened to update the images https://github.com/mozilla/blurts-server/pull/2799
- **Optimize DOM queries**
That’s a very good point and a desirable optimization. I created [MNTOR-1139](https://mozilla-hub.atlassian.net/browse/MNTOR-1139) for a follow-up.

**Pull request merged**
